model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/remotestore/RemoteCacheSessionListener.java

```diff
@@ -156,18 +183,35 @@ protected void replaceRemoteEntityInCache(K key, long eventVersion) {
             }
             SessionEntity remoteSession = remoteSessionVersioned.getValue().getEntity();
 
-            logger.debugf("Read session entity from the remote cache: %s . replaceRetries=%d", remoteSession.toString(), replaceRetries);
+            logger.tracef("Read session entity from the remote cache: %s . replaceRetries=%d", remoteSession, replaceRetries);
 
             SessionEntityWrapper<V> sessionWrapper = remoteSession.mergeRemoteEntityWithLocalEntity(localEntityWrapper);
 
+            RealmModel realm = realmRef.get();
+            if (realm == null) {
+
+                KeycloakModelUtils.runJobInTransaction(sessionFactory, (session -> {
+
+                    realmRef.set(session.realms().getRealm(sessionWrapper.getEntity().getRealmId()));
+
+                }));
+
+                realm = realmRef.get();
+            }
+
+            long lifespanMs = lifespanMsLoader.apply(realm, sessionWrapper.getEntity());
+            long maxIdleTimeMs = maxIdleTimeMsLoader.apply(realm, sessionWrapper.getEntity());
```

Comment:
Could determining `lifespanMs` and `maxIdleTimeMs` be returned back into the `runJobInTransaction` function? The current state may not work if e.g. the `realm` is only a partially cached object that does not have lifespans defined and delegates the lookup lazily only when queried; however this would fail since the lazy loading needs an active transaction and the transaction currently ends before this lookup.

Disclaimer: We don't have it in the current codebase but it is possible, even with the current store, and it is one of the discussed optimization for the new store.

Reply:
@hmlnarik Thanks for clarifying this. I did not consider this for the new store.

So the points to consider:
- lifespanMsLoader and maxIdleTimeMsLoader needs to be called within the transaction due the reason you mentioned
- Obtaining of RealmModel can't be easily moved outside of `do-while` to the top. As the ID of the realm is retrieved from the session, which is loaded inside the `do` block. With my latest change, the load of the RealmModel will be done only once (due the realmRef), however it won't work due the previous point
- Computing of `lifespanMs` and `maxIdleTimeMs` depends on the realm, but also on the sessionEntity and it's "lastSessionRefresh" (for UserSEssion) or "timestamp" (ClientSession). And this timestamps can change. For example in first iteration of `do` the session may have timestamp 1000, but in the second iteration, the timestamp can be 2000 (as there was an update from concurrent transaction, which updated the timestamp and which then caused this conflict and the fact that second iteration of `do` was needed).

Regarding this, I see the options are:
1) Wrap the whole `do-while` into the `runJobInTransaction` block
2) Revert to the behaviour as was before this change.
It seems the (1) is better due the fact that (2) can mean multiple transactions as you pointed (even if this shouldn't happen too often as it is not big chance that same session is updated concurrently, but it is indeed possibility).

Any better idea?

Reply:
> Computing of lifespanMs and maxIdleTimeMs depends on the realm, but also on the sessionEntity and it's "lastSessionRefresh" (for UserSEssion) or "timestamp" (ClientSession). And this timestamps can change. For example in first iteration of do the session may have timestamp 1000, but in the second iteration, the timestamp can be 2000 (as there was an update from concurrent transaction, which updated the timestamp and which then caused this conflict and the fact that second iteration of do was needed).

I see, and you're right. If they have changed, then we need to start a separate transaction, otherwise the single transaction could always see the stale values and never succeed. For that reason alone, I think we should follow the Option 2. Apologies for confusion.

Reply:
@hmlnarik Thanks! I've updated the PR to use the option 2.